File: packages/protocol-ts/src/negotiation.ts
import { timingSafeEqual } from "node:crypto";
import { create } from "@bufbuild/protobuf";

import {
  type ProtocolInfo,
  type ProtocolLimits,
  ProtocolLimitsSchema,
  type ProtocolSelectionSnapshot,
} from "./generated/loop/v1/common_pb.js";
import { protocol_selection_sha256 } from "./job.js";

const MAX_PACKAGES = 16;
const MAX_FEATURES = 256;
const MAX_NAME_BYTES = 128;
const MAX_BUILD_VERSION_BYTES = 128;
const MAX_UNARY_BYTES = 4_194_304n;
const MAX_STREAM_EVENT_BYTES = 1_048_576n;
const MAX_CANONICAL_AST_BYTES = 262_144n;
const MAX_AST_NODES = 4_096;
const MAX_AST_DEPTH = 64;
const MAX_PAGE_RECORDS = 500;
const MAX_IDENTITY_BYTES = 128;
const MAX_ARTIFACT_URI_BYTES = 2_048;
const encoder = new TextEncoder();

export type ProtocolNegotiationCode =
  | "invalid_protocol_info"
  | "invalid_requirement"
  | "unsupported_major"
  | "unsupported_required_feature"
  | "unsupported_feature"
  | "incompatible_limits"
  | "invalid_selection"
  | "unavailable_build"
  | "unavailable_descriptor";

export class ProtocolNegotiationError extends Error {
  public override readonly name = "ProtocolNegotiationError";

  public constructor(public readonly code: ProtocolNegotiationCode) {
    super(code);
  }
}

export interface ProtocolBuildIdentity {
  readonly buildVersion: string;
  readonly buildSha256: Uint8Array;
}

export interface NegotiatedProtocol {
  readonly selectedPackage: string;
  readonly enabledFeatures: readonly string[];
  readonly effectiveLimits: ProtocolLimits;
  readonly localBuild: ProtocolBuildIdentity;
  readonly peerBuild: ProtocolBuildIdentity;
}

/** Validate the closed availability projection advertised by one deployed peer. */
export function validate_protocol_info(info: ProtocolInfo): void {
  if (
    info.supportedPackages.length < 1 ||
    info.supportedPackages.length > MAX_PACKAGES ||
    !is_sorted_unique(info.supportedPackages) ||
    info.supportedPackages.some((value) => !is_protocol_package(value)) ||
    info.features.length > MAX_FEATURES ||
    !is_sorted_unique(info.features) ||
    info.features.some((value) => !is_protocol_feature(value)) ||
    !is_valid_limits(info.limits) ||
    !is_build_version(info.buildVersion) ||
    info.buildSha256?.value.byteLength !== 32
  ) {
    throw new ProtocolNegotiationError("invalid_protocol_info");
  }
}

/** Negotiate one explicitly required service major before mutable or paid work. */
export function negotiate_protocol_availability(
  local: ProtocolInfo,
  peer: ProtocolInfo,
  requiredPackage: string,
  requiredFeatures: readonly string[],
): Readonly<NegotiatedProtocol> {
  validate_protocol_info(local);
  validate_protocol_info(peer);
  validate_requirement(requiredPackage, requiredFeatures);
  if (
    !local.supportedPackages.includes(requiredPackage) ||
    !peer.supportedPackages.includes(requiredPackage)
  ) {
    throw new ProtocolNegotiationError("unsupported_major");
  }
  const peerFeatures = new Set(peer.features);
  const enabledFeatures = local.features.filter((feature) => peerFeatures.has(feature));
  if (requiredFeatures.some((required) => !enabledFeatures.includes(required))) {
    throw new ProtocolNegotiationError("unsupported_required_feature");
  }

  return Object.freeze({
    selectedPackage: requiredPackage,
    enabledFeatures: Object.freeze(enabledFeatures),
    effectiveLimits: minimum_limits(
      require_limits(local.limits, "invalid_protocol_info"),
      require_limits(peer.limits, "invalid_protocol_info"),
    ),
    localBuild: protocol_info_build(local),
    peerBuild: protocol_info_build(peer),
  });
}

/** Prove a pinned selection remains executable from local immutable content. */
export function validate_selection_availability(
  selection: ProtocolSelectionSnapshot,
  local: ProtocolInfo,
  retainedBuilds: readonly ProtocolBuildIdentity[],
  availableSchemaDescriptors: readonly Uint8Array[],
  requiredPackage: string,
  requiredFeatures: readonly string[],
): void {
  validate_protocol_info(local);
  validate_requirement(requiredPackage, requiredFeatures);
  let computed: Uint8Array;
  try {
    computed = protocol_selection_sha256(selection);
  } catch {
    throw new ProtocolNegotiationError("invalid_selection");
  }
  if (
    selection.selectionSha256?.value.byteLength !== 32 ||
    !equal_bytes(selection.selectionSha256.value, computed)
  ) {
    throw new ProtocolNegotiationError("invalid_selection");
  }
  if (
    selection.selectedPackage !== requiredPackage ||
    !local.supportedPackages.includes(selection.selectedPackage)
  ) {
    throw new ProtocolNegotiationError("unsupported_major");
  }
  if (requiredFeatures.some((required) => !selection.enabledFeatures.includes(required))) {
    throw new ProtocolNegotiationError("unsupported_required_feature");
  }
  if (selection.enabledFeatures.some((feature) => !local.features.includes(feature))) {
    throw new ProtocolNegotiationError("unsupported_feature");
  }
  const selectedLimits = require_limits(selection.effectiveLimits, "invalid_selection");
  const localLimits = require_limits(local.limits, "invalid_protocol_info");
  if (!limits_fit(selectedLimits, localLimits)) {
    throw new ProtocolNegotiationError("incompatible_limits");
  }

  const serverBuild = selection_build(
    selection.serverBuildVersion,
    selection.serverBuildSha256?.value,
  );
  const clientBuild = selection_build(
    selection.clientBuildVersion,
    selection.clientBuildSha256?.value,
  );
  if (
    !build_is_available(serverBuild, local, retainedBuilds) ||
    !build_is_available(clientBuild, local, retainedBuilds)
  ) {
    throw new ProtocolNegotiationError("unavailable_build");
  }
  const descriptor = selection.schemaDescriptorSha256?.value;
  if (
    descriptor?.byteLength !== 32 ||
    !availableSchemaDescriptors.some((available) => equal_bytes(available, descriptor))
  ) {
    throw new ProtocolNegotiationError(
      descriptor?.byteLength === 32 ? "unavailable_descriptor" : "invalid_selection",
    );
  }
}

function validate_requirement(requiredPackage: string, requiredFeatures: readonly string[]): void {
  if (
    !is_protocol_package(requiredPackage) ||
    requiredFeatures.length > MAX_FEATURES ||
    !is_sorted_unique(requiredFeatures) ||
    requiredFeatures.some((feature) => !is_protocol_feature(feature))
  ) {
    throw new ProtocolNegotiationError("invalid_requirement");
  }
}

function protocol_info_build(info: ProtocolInfo): ProtocolBuildIdentity {
  const digest = info.buildSha256?.value;
  if (digest?.byteLength !== 32) throw new ProtocolNegotiationError("invalid_protocol_info");
  return Object.freeze({ buildVersion: info.buildVersion, buildSha256: digest.slice() });
}

function selection_build(version: string, digest: Uint8Array | undefined): ProtocolBuildIdentity {
  if (digest?.byteLength !== 32) throw new ProtocolNegotiationError("invalid_selection");
  return { buildVersion: version, buildSha256: digest };
}

function build_is_available(
  requested: ProtocolBuildIdentity,
  local: ProtocolInfo,
  retained: readonly ProtocolBuildIdentity[],
): boolean {
  const current = protocol_info_build(local);
  return (
    build_identity_equal(requested, current) ||
    retained.some((available) => build_identity_equal(requested, available))
  );
}

function build_identity_equal(left: ProtocolBuildIdentity, right: ProtocolBuildIdentity): boolean {
  return (
    left.buildVersion === right.buildVersion && equal_bytes(left.buildSha256, right.buildSha256)
  );
}

function minimum_limits(left: ProtocolLimits, right: ProtocolLimits): ProtocolLimits {
  return create(ProtocolLimitsSchema, {
    maximumUnaryBytes:
      left.maximumUnaryBytes < right.maximumUnaryBytes
        ? left.maximumUnaryBytes
        : right.maximumUnaryBytes,
    maximumStreamEventBytes:
      left.maximumStreamEventBytes < right.maximumStreamEventBytes
        ? left.maximumStreamEventBytes
        : right.maximumStreamEventBytes,
    maximumCanonicalAstBytes:
      left.maximumCanonicalAstBytes < right.maximumCanonicalAstBytes
        ? left.maximumCanonicalAstBytes
        : right.maximumCanonicalAstBytes,
    maximumAstNodes: Math.min(left.maximumAstNodes, right.maximumAstNodes),
    maximumAstDepth: Math.min(left.maximumAstDepth, right.maximumAstDepth),
    maximumPageRecords: Math.min(left.maximumPageRecords, right.maximumPageRecords),
    maximumIdentityBytes: Math.min(left.maximumIdentityBytes, right.maximumIdentityBytes),
    maximumArtifactUriBytes: Math.min(left.maximumArtifactUriBytes, right.maximumArtifactUriBytes),
  });
}

function limits_fit(selected: ProtocolLimits, local: ProtocolLimits): boolean {
  return (
    selected.maximumUnaryBytes <= local.maximumUnaryBytes &&
    selected.maximumStreamEventBytes <= local.maximumStreamEventBytes &&
    selected.maximumCanonicalAstBytes <= local.maximumCanonicalAstBytes &&
    selected.maximumAstNodes <= local.maximumAstNodes &&
    selected.maximumAstDepth <= local.maximumAstDepth &&
    selected.maximumPageRecords <= local.maximumPageRecords &&
    selected.maximumIdentityBytes <= local.maximumIdentityBytes &&
    selected.maximumArtifactUriBytes <= local.maximumArtifactUriBytes
  );
}

function is_valid_limits(limits: ProtocolLimits | undefined): boolean {
  return (
    limits !== undefined &&
    limits.maximumUnaryBytes >= 1n &&
    limits.maximumUnaryBytes <= MAX_UNARY_BYTES &&
    limits.maximumStreamEventBytes >= 1n &&
    limits.maximumStreamEventBytes <= MAX_STREAM_EVENT_BYTES &&
    limits.maximumCanonicalAstBytes >= 1n &&
    limits.maximumCanonicalAstBytes <= MAX_CANONICAL_AST_BYTES &&
    limits.maximumAstNodes >= 1 &&
    limits.maximumAstNodes <= MAX_AST_NODES &&
    limits.maximumAstDepth >= 1 &&
    limits.maximumAstDepth <= MAX_AST_DEPTH &&
    limits.maximumPageRecords >= 1 &&
    limits.maximumPageRecords <= MAX_PAGE_RECORDS &&
    limits.maximumIdentityBytes >= 1 &&
    limits.maximumIdentityBytes <= MAX_IDENTITY_BYTES &&
    limits.maximumArtifactUriBytes >= 1 &&
    limits.maximumArtifactUriBytes <= MAX_ARTIFACT_URI_BYTES &&
    limits.maximumCanonicalAstBytes <= limits.maximumUnaryBytes &&
    BigInt(limits.maximumIdentityBytes) <= limits.maximumUnaryBytes &&
    BigInt(limits.maximumArtifactUriBytes) <= limits.maximumUnaryBytes
  );
}

function require_limits(
  limits: ProtocolLimits | undefined,
  code: "invalid_protocol_info" | "invalid_selection",
): ProtocolLimits {
  if (limits === undefined) throw new ProtocolNegotiationError(code);
  return limits;
}

function is_sorted_unique(values: readonly string[]): boolean {
  return values.every((value, index) => index === 0 || (values[index - 1] as string) < value);
}

function is_protocol_package(value: string): boolean {
  return (
    encoder.encode(value).byteLength <= MAX_NAME_BYTES &&
    /^[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*)*\.v[1-9][0-9]*$/.test(value)
  );
}

function is_protocol_feature(value: string): boolean {
  return (
    encoder.encode(value).byteLength <= MAX_NAME_BYTES &&
    /^[a-z0-9](?:[a-z0-9-]*[a-z0-9])?(?:\.[a-z0-9](?:[a-z0-9-]*[a-z0-9])?)+$/.test(value)
  );
}

function is_build_version(value: string): boolean {
  return (
    encoder.encode(value).byteLength <= MAX_BUILD_VERSION_BYTES &&
    /^[A-Za-z0-9][A-Za-z0-9.+_-]*$/.test(value)
  );
}

function equal_bytes(left: Uint8Array, right: Uint8Array): boolean {
  return left.byteLength === right.byteLength && timingSafeEqual(left, right);
}

File: python/loop_protocol/src/loop_protocol/negotiation.py
"""Pure protocol availability and negotiation checks for preflight gates."""

from __future__ import annotations

import hmac
import re
from collections.abc import Sequence
from dataclasses import dataclass
from enum import StrEnum
from itertools import pairwise

from loop.v1 import common_pb2

from .job import JobValidationError, protocol_selection_sha256

_MAX_PACKAGES = 16
_MAX_FEATURES = 256
_MAX_NAME_BYTES = 128
_MAX_BUILD_VERSION_BYTES = 128
_MAX_UNARY_BYTES = 4_194_304
_MAX_STREAM_EVENT_BYTES = 1_048_576
_MAX_CANONICAL_AST_BYTES = 262_144
_MAX_AST_NODES = 4_096
_MAX_AST_DEPTH = 64
_MAX_PAGE_RECORDS = 500
_MAX_IDENTITY_BYTES = 128
_MAX_ARTIFACT_URI_BYTES = 2_048
_PROTOCOL_PACKAGE_RE = re.compile(
    r"^[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*)*\.v[1-9][0-9]*$", re.ASCII
)
_PROTOCOL_FEATURE_RE = re.compile(
    r"^[a-z0-9](?:[a-z0-9-]*[a-z0-9])?(?:\.[a-z0-9](?:[a-z0-9-]*[a-z0-9])?)+$",
    re.ASCII,
)
_BUILD_VERSION_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9.+_-]*$", re.ASCII)


class ProtocolNegotiationCode(StrEnum):
    """Stable fail-closed preflight result codes shared by all runtimes."""

    INVALID_PROTOCOL_INFO = "invalid_protocol_info"
    INVALID_REQUIREMENT = "invalid_requirement"
    UNSUPPORTED_MAJOR = "unsupported_major"
    UNSUPPORTED_REQUIRED_FEATURE = "unsupported_required_feature"
    UNSUPPORTED_FEATURE = "unsupported_feature"
    INCOMPATIBLE_LIMITS = "incompatible_limits"
    INVALID_SELECTION = "invalid_selection"
    UNAVAILABLE_BUILD = "unavailable_build"
    UNAVAILABLE_DESCRIPTOR = "unavailable_descriptor"


class ProtocolNegotiationError(ValueError):
    """Protocol availability cannot satisfy a preflight requirement."""

    def __init__(self, code: ProtocolNegotiationCode) -> None:
        self.code = code
        super().__init__(code.value)


@dataclass(frozen=True, slots=True)
class ProtocolBuildIdentity:
    build_version: str
    build_sha256: bytes


@dataclass(frozen=True, slots=True)
class NegotiatedProtocol:
    selected_package: str
    enabled_features: tuple[str, ...]
    effective_limits: common_pb2.ProtocolLimits
    local_build: ProtocolBuildIdentity
    peer_build: ProtocolBuildIdentity


def validate_protocol_info(info: common_pb2.ProtocolInfo) -> None:
    """Validate the closed availability projection advertised by one peer."""

    valid = (
        1 <= len(info.supported_packages) <= _MAX_PACKAGES
        and _is_sorted_unique(info.supported_packages)
        and all(_is_protocol_package(value) for value in info.supported_packages)
        and len(info.features) <= _MAX_FEATURES
        and _is_sorted_unique(info.features)
        and all(_is_protocol_feature(value) for value in info.features)
        and info.HasField("limits")
        and _is_valid_limits(info.limits)
        and _is_build_version(info.build_version)
        and info.HasField("build_sha256")
        and len(info.build_sha256.value) == 32
    )
    if not valid:
        raise ProtocolNegotiationError(ProtocolNegotiationCode.INVALID_PROTOCOL_INFO)


def negotiate_protocol_availability(
    local: common_pb2.ProtocolInfo,
    peer: common_pb2.ProtocolInfo,
    required_package: str,
    required_features: Sequence[str],
) -> NegotiatedProtocol:
    """Negotiate one required service major before mutable or paid work."""

    validate_protocol_info(local)
    validate_protocol_info(peer)
    _validate_requirement(required_package, required_features)
    if (
        required_package not in local.supported_packages
        or required_package not in peer.supported_packages
    ):
        raise ProtocolNegotiationError(ProtocolNegotiationCode.UNSUPPORTED_MAJOR)

    peer_features = frozenset(peer.features)
    enabled_features = tuple(feature for feature in local.features if feature in peer_features)
    if any(feature not in enabled_features for feature in required_features):
        raise ProtocolNegotiationError(ProtocolNegotiationCode.UNSUPPORTED_REQUIRED_FEATURE)

    return NegotiatedProtocol(
        selected_package=required_package,
        enabled_features=enabled_features,
        effective_limits=_minimum_limits(local.limits, peer.limits),
        local_build=_protocol_info_build(local),
        peer_build=_protocol_info_build(peer),
    )


def validate_protocol_selection_availability(
    selection: common_pb2.ProtocolSelectionSnapshot,
    local: common_pb2.ProtocolInfo,
    retained_builds: Sequence[ProtocolBuildIdentity],
    available_schema_descriptors: Sequence[bytes],
    required_package: str,
    required_features: Sequence[str],
) -> None:
    """Prove a pinned selection remains executable from local immutable content."""

    validate_protocol_info(local)
    _validate_requirement(required_package, required_features)
    try:
        computed = protocol_selection_sha256(selection)
    except JobValidationError as error:
        raise ProtocolNegotiationError(ProtocolNegotiationCode.INVALID_SELECTION) from error
    if (
        not selection.HasField("selection_sha256")
        or len(selection.selection_sha256.value) != 32
        or not hmac.compare_digest(selection.selection_sha256.value, computed)
    ):
        raise ProtocolNegotiationError(ProtocolNegotiationCode.INVALID_SELECTION)
    if (
        selection.selected_package != required_package
        or selection.selected_package not in local.supported_packages
    ):
        raise ProtocolNegotiationError(ProtocolNegotiationCode.UNSUPPORTED_MAJOR)
    if any(feature not in selection.enabled_features for feature in required_features):
        raise ProtocolNegotiationError(ProtocolNegotiationCode.UNSUPPORTED_REQUIRED_FEATURE)
    if any(feature not in local.features for feature in selection.enabled_features):
        raise ProtocolNegotiationError(ProtocolNegotiationCode.UNSUPPORTED_FEATURE)
    if not _limits_fit(selection.effective_limits, local.limits):
        raise ProtocolNegotiationError(ProtocolNegotiationCode.INCOMPATIBLE_LIMITS)

    server_build = _selection_build(
        selection.server_build_version,
        selection.server_build_sha256.value,
    )
    client_build = _selection_build(
        selection.client_build_version,
        selection.client_build_sha256.value,
    )
    if not _build_is_available(server_build, local, retained_builds) or not _build_is_available(
        client_build, local, retained_builds
    ):
        raise ProtocolNegotiationError(ProtocolNegotiationCode.UNAVAILABLE_BUILD)
    descriptor = selection.schema_descriptor_sha256.value
    if not any(
        hmac.compare_digest(descriptor, available) for available in available_schema_descriptors
    ):
        raise ProtocolNegotiationError(ProtocolNegotiationCode.UNAVAILABLE_DESCRIPTOR)


def _validate_requirement(required_package: str, required_features: Sequence[str]) -> None:
    valid = (
        _is_protocol_package(required_package)
        and len(required_features) <= _MAX_FEATURES
        and _is_sorted_unique(required_features)
        and all(_is_protocol_feature(value) for value in required_features)
    )
    if not valid:
        raise ProtocolNegotiationError(ProtocolNegotiationCode.INVALID_REQUIREMENT)


def _protocol_info_build(info: common_pb2.ProtocolInfo) -> ProtocolBuildIdentity:
    return ProtocolBuildIdentity(info.build_version, bytes(info.build_sha256.value))


def _selection_build(version: str, digest: bytes) -> ProtocolBuildIdentity:
    if not _is_build_version(version) or len(digest) != 32:
        raise ProtocolNegotiationError(ProtocolNegotiationCode.INVALID_SELECTION)
    return ProtocolBuildIdentity(version, bytes(digest))


def _build_is_available(
    requested: ProtocolBuildIdentity,
    local: common_pb2.ProtocolInfo,
    retained: Sequence[ProtocolBuildIdentity],
) -> bool:
    return _build_identity_equal(requested, _protocol_info_build(local)) or any(
        _build_identity_equal(requested, available) for available in retained
    )


def _build_identity_equal(left: ProtocolBuildIdentity, right: ProtocolBuildIdentity) -> bool:
    return left.build_version == right.build_version and hmac.compare_digest(
        left.build_sha256, right.build_sha256
    )


def _minimum_limits(
    left: common_pb2.ProtocolLimits, right: common_pb2.ProtocolLimits
) -> common_pb2.ProtocolLimits:
    return common_pb2.ProtocolLimits(
        maximum_unary_bytes=min(left.maximum_unary_bytes, right.maximum_unary_bytes),
        maximum_stream_event_bytes=min(
            left.maximum_stream_event_bytes, right.maximum_stream_event_bytes
        ),
        maximum_canonical_ast_bytes=min(
            left.maximum_canonical_ast_bytes, right.maximum_canonical_ast_bytes
        ),
        maximum_ast_nodes=min(left.maximum_ast_nodes, right.maximum_ast_nodes),
        maximum_ast_depth=min(left.maximum_ast_depth, right.maximum_ast_depth),
        maximum_page_records=min(left.maximum_page_records, right.maximum_page_records),
        maximum_identity_bytes=min(left.maximum_identity_bytes, right.maximum_identity_bytes),
        maximum_artifact_uri_bytes=min(
            left.maximum_artifact_uri_bytes, right.maximum_artifact_uri_bytes
        ),
    )


def _limits_fit(selected: common_pb2.ProtocolLimits, local: common_pb2.ProtocolLimits) -> bool:
    return (
        selected.maximum_unary_bytes <= local.maximum_unary_bytes
        and selected.maximum_stream_event_bytes <= local.maximum_stream_event_bytes
        and selected.maximum_canonical_ast_bytes <= local.maximum_canonical_ast_bytes
        and selected.maximum_ast_nodes <= local.maximum_ast_nodes
        and selected.maximum_ast_depth <= local.maximum_ast_depth
        and selected.maximum_page_records <= local.maximum_page_records
        and selected.maximum_identity_bytes <= local.maximum_identity_bytes
        and selected.maximum_artifact_uri_bytes <= local.maximum_artifact_uri_bytes
    )


def _is_valid_limits(limits: common_pb2.ProtocolLimits) -> bool:
    return (
        1 <= limits.maximum_unary_bytes <= _MAX_UNARY_BYTES
        and 1 <= limits.maximum_stream_event_bytes <= _MAX_STREAM_EVENT_BYTES
        and 1 <= limits.maximum_canonical_ast_bytes <= _MAX_CANONICAL_AST_BYTES
        and 1 <= limits.maximum_ast_nodes <= _MAX_AST_NODES
        and 1 <= limits.maximum_ast_depth <= _MAX_AST_DEPTH
        and 1 <= limits.maximum_page_records <= _MAX_PAGE_RECORDS
        and 1 <= limits.maximum_identity_bytes <= _MAX_IDENTITY_BYTES
        and 1 <= limits.maximum_artifact_uri_bytes <= _MAX_ARTIFACT_URI_BYTES
        and limits.maximum_canonical_ast_bytes <= limits.maximum_unary_bytes
        and limits.maximum_identity_bytes <= limits.maximum_unary_bytes
        and limits.maximum_artifact_uri_bytes <= limits.maximum_unary_bytes
    )


def _is_sorted_unique(values: Sequence[str]) -> bool:
    return all(left < right for left, right in pairwise(values))


def _is_protocol_package(value: str) -> bool:
    return len(value.encode("utf-8")) <= _MAX_NAME_BYTES and bool(
        _PROTOCOL_PACKAGE_RE.fullmatch(value)
    )


def _is_protocol_feature(value: str) -> bool:
    return len(value.encode("utf-8")) <= _MAX_NAME_BYTES and bool(
        _PROTOCOL_FEATURE_RE.fullmatch(value)
    )


def _is_build_version(value: str) -> bool:
    return len(value.encode("utf-8")) <= _MAX_BUILD_VERSION_BYTES and bool(
        _BUILD_VERSION_RE.fullmatch(value)
    )
